pinot-core/src/main/java/org/apache/pinot/core/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.upsert;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.pinot.core.realtime.impl.ThreadSafeMutableRoaringBitmap;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Manages the upsert metadata per partition.
+ */
+@ThreadSafe
+public class PartitionUpsertMetadataManager {
```

Comment:
I don't quite get this. Do you suggest other implementations of the Upsert Metadata Manager by introducing the interface? It's unclear to how this helps the null checking.

Reply:
I mean for a new component like this it will be great if they are implemented as interface -> concrete class. This provides multiple benefits:

1. it allows people to use different implementations of the same upsert logics. Think of someone who wants to use a key/value mapping backed by rocksdb, it will be much easier to swap this part out if it is backed by an interface.

2. it allows us to provide a dummy implementation of MetadataManager for append-only segmentImpl. For example, this could be one `PartitionUpsertMetadataManager` and `PartitionAppendMetadataManager` both extends `PartitionMetadataManager` interface. This way it would allow us to create a concrete class based on whether we are dealing with upsert/append table and avoid the branching conditions for append/upsert logics

Reply:
I see.

I think we can do the abstraction and add the interface when we come to it. For now, I feel it's a bit too early to add this generalization. You know, I'm a bit influenced by [this philosophy]( https://wiki.c2.com/?PrematureGeneralizationIsEvil) :)